app/src/main/java/com/tari/android/wallet/infrastructure/security/encryption/SymmetricEncryptionAlgorithm.kt

```diff
@@ -0,0 +1,56 @@
+package com.tari.android.wallet.infrastructure.security.encryption
+
+import java.io.InputStream
+import java.io.OutputStream
+import java.security.MessageDigest
+import javax.crypto.Cipher
+import javax.crypto.CipherOutputStream
+import javax.crypto.spec.SecretKeySpec
+
+interface SymmetricEncryptionAlgorithm {
+
+    fun encrypt(key: CharArray, isProvider: () -> InputStream, osProvider: () -> OutputStream)
+
+    companion object {
+        fun aes(): SymmetricEncryptionAlgorithm = AdvancedEncryptionStandard()
+    }
+
+}
+
+private class AdvancedEncryptionStandard(
+    private val digest: MessageDigest = MessageDigest.getInstance(DIGEST_ALGORITHM)
+) : SymmetricEncryptionAlgorithm {
+
+    // TODO add salt?
+    // TODO add padding and initialization vector?
+    override fun encrypt(
+        key: CharArray,
+        isProvider: () -> InputStream,
```

Comment:
`isProvider` => `inputStreamProvider`

Reply:
Can't you pass the stream directly? You run the provider immediately, so I'm not sure if the indirection is justified.

Reply:
it's for RAII ( https://en.wikipedia.org/wiki/Resource_acquisition_is_initialization )

there is a possibility that abstraction permits that the consumer won't use either of streams, so the consumer will obtain a stream through creation only is he needs